### PR TITLE
[future] Undo some recent torch::utils::Future api changes

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36220 [future] Undo some recent torch::util::future api changes**

The torch::utils::Future change from yesterday may have introduced a reference cycle,
leading to OOM on PS. This change reverts the lambda  capture changes with
torch::utils::Future until we can analyze further.

Differential Revision: [D20918904](https://our.internmc.facebook.com/intern/diff/D20918904/)

### CODE REVIEW

**Review by mrshenli — APPROVED** [review]
For reference of release note, this is reverting part of the logic in #35849. 

Stamp to unblock. Please wait for tests before landing. Thanks!

**Comment by jjlilley** [comment]
Thanks for reviewing this! Waiting for the tests.

Discovered that ctr_mobile_feed was working about an hour before #35849 went in last night, then started OOMing after.

As a data point,  when trying to repro this morning:
  - I can repro this OOM with ctr_mobile_feed within ~10 min
  - whereas with this change applied, no problems running 25 min now.

**Pull request closed without merging**
